src/daemon/metric.h

```diff
@@ -0,0 +1,185 @@
+ *   Manoj Srivastava <srivasta at google.com>
+ **/
+
+#ifndef METRIC_H
+#define METRIC_H 1
+
+#include "utils/metadata/meta_data.h"
+#include "utils/strbuf/strbuf.h"
+#include "utils_time.h"
+
+#define VALUE_TYPE_GAUGE 1
+#define VALUE_TYPE_DERIVE 2
+
+typedef enum {
+  METRIC_TYPE_COUNTER = 0,
+  METRIC_TYPE_GAUGE = 1,
+  METRIC_TYPE_UNTYPED = 2,
+} metric_type_t;
+
+typedef uint64_t counter_t;
+typedef double gauge_t;
+typedef int64_t derive_t;
+
+union value_u {
+  counter_t counter;
+  gauge_t gauge;
+  derive_t derive;
+};
+typedef union value_u value_t;
+
+/* value_marshal_text prints a text representation of v to buf. */
+int value_marshal_text(strbuf_t *buf, value_t v, metric_type_t type);
+
+/*
+ * Labels
+ */
+/* label_pair_t represents a label, i.e. a key/value pair. */
+typedef struct {
+  char *name;
+  char *value;
+} label_pair_t;
+
+/* label_t represents a key/value pair. It is similar to label_pair_t, except
+ * that is has const fields. label_t is used in function arguments to prevent
+ * the called function from modifying its argument. Internally labels are
+ * stored as label_pair_t to allow modification, e.g. by targets in the "filter
+ * chain". */
+typedef struct {
+  char const *name;
+  char const *value;
+} label_t;
+
+/* label_set_t is a sorted set of labels. */
+typedef struct {
+  label_pair_t *ptr;
+  size_t num;
+} label_set_t;
+
+/*
+ * Metric
+ */
+/* forward declaration since metric_family_t and metric_t refer to each other.
+ */
+struct metric_family_s;
+typedef struct metric_family_s metric_family_t;
+
+/* metric_t is a metric inside a metric family. */
+typedef struct {
+  metric_family_t *family; /* for family->name and family->type */
+
+  label_set_t label;
+  value_t value;
+  cdtime_t time; /* TODO(octo): use ms or µs instead? */
+  cdtime_t interval;
+  /* TODO(octo): target labels */
+  meta_data_t *meta;
+} metric_t;
+
+/* metric_identity writes the identity of the metric "m" to "buf". An example
+ * string is:
+ *
+ *   "http_requests_total{method=\"post\",code=\"200\"}"
+ */
+int metric_identity(strbuf_t *buf, metric_t const *m);
+
+/* metric_label_set adds or updates a label to/in the label set.
+ * If "value" is NULL or the empty string, the label is removed. Removing a
+ * label that does not exist is *not* an error. */
+int metric_label_set(metric_t *m, char const *name, char const *value);
+
+/* metric_label_get efficiently looks up and returns the value of the "name"
+ * label. If a label does not exist, NULL is returned and errno is set to
+ * ENOENT. The returned pointer may not be valid after a subsequent call to
+ * "metric_label_set". */
+char const *metric_label_get(metric_t const *m, char const *name);
+
+/* metric_reset frees all labels and meta data stored in the metric and resets
+ * the metric to zero. */
+int metric_reset(metric_t *m);
+
+/* metric_marshal_text writes a text representation of "m" to the buffer "buf".
+ * An example output is:
```

Comment:
This one is a bit trickier, because it doesn't implement all that OpenMetrics / Prometheus specify, just what `metric_marshal_text` produces.

`metric_identity` is used, e.g. to look up the metric in the cache (see `src/daemon/utils_cache.c`). Arguably `metric_marshal_text` and `metric_family_unmarshal_text` could be moved to an OpenMetrics / Prometheus specific plugin. What do you think?

Reply:
I like that, since that leaves open the possibility of marshaling/unmarshaling to a different wire protocol in the future, if needed, without increasing complexity a whole lot. And we can add a comment in the documentation in the plugin poninting to the spec, and stating that only the relevant bitsof the spec are implemented here

Reply:
I removed these two functions – well, partially. I added `metric_parse_identity()`, which is used e.g. when a metric expires: we get the string representation from the cache and need to convert it to a `metric_t*` or `metric_family_t*`.